File: us_payroll/custom_script/leave_allocation.js
frappe.ui.form.on("Leave Allocation", {
	onload: function (frm) {
		frm.trigger("set_fiscal_year_dates");
	},

	set_fiscal_year_dates: function (frm) {
		if (!frm.doc.from_date)
			frm.set_value(
				"from_date",
				erpnext.utils.get_fiscal_year(frappe.datetime.get_today(), true)[1]
			);
		if (!frm.doc.to_date)
			frm.set_value(
				"to_date",
				erpnext.utils.get_fiscal_year(frappe.datetime.get_today(), true)[2]
			);
	},

	new_leaves_allocated: function (frm) {
		if (!frm.__islocal && frm.doc.new_leaves_allocated) {
			frappe.call({
				method: "us_payroll.custom_script.leave_allocation.get_old_leave_allocation_amount",
				args: {
					employee: frm.doc.employee,
					amount: parseFloat(frm.doc.new_leaves_allocated),
				},
				callback: function (r) {
					let res = r.message;
					if (res.status === "exists_different_amount") {
						frappe.confirm(
							`Previous allocation was ${res.existing_amount} for employee ${frm.doc.employee}. Do you want to update it to ${res.new_amount}?`,
							function () {
								show_popup_for_reason(frm, res.existing_amount, res.new_amount);
							},
							function () {
								frm.set_value(
									"new_leaves_allocated",
									parseFloat(res.existing_amount)
								);
								frappe.msgprint("Operation cancelled.");
							}
						);
					} else if (res.status === "created") {
						frm.save();
					}
				},
			});
		}
	},
});

function show_popup_for_reason(frm, existing_amount, new_amount) {
	let dialog = new frappe.ui.Dialog({
		title: "Reason for Change",
		fields: [
			{
				label: "Reason",
				fieldname: "reason",
				fieldtype: "Small Text",
				reqd: 1,
			},
			{
				label: "New amount",
				fieldname: "new_amount",
				fieldtype: "Currency",
				default: new_amount,
				read_only: 1,
			},
		],
		primary_action_label: "Submit",
		primary_action(values) {
			frappe.call({
				method: "us_payroll.custom_script.leave_allocation.update_allocated_leaves",
				args: {
					employee: frm.doc.employee,
					amount: new_amount,
					reason: values.reason,
					existing_amount: existing_amount,
				},
				callback: function (res) {
					frm.set_value("custom_allocation_change_reason", values.reason);
					frm.set_value("custom_previous_leaves_allocated", existing_amount);
					frm.set_value("new_leaves_allocated", parseFloat(values.new_amount));

					frappe.msgprint("Leave allocation updated successfully.");
					dialog.hide();
					frm.reload_doc();
				},
			});
		},
	});
	dialog.show();
}

File: us_payroll/custom_script/leave_allocation.py
import json

import frappe
from frappe.utils import nowdate


def validate(doc, method):
	pass


def before_insert(doc, method):
	pass


def after_insert(doc, method):
	set_transaction_date(doc)


def set_transaction_date(doc):
	if not doc.custom_transaction_date:
		doc.custom_transaction_date = nowdate()


@frappe.whitelist()
def get_old_leave_allocation_amount(employee: str, amount: float):
	fy_info = get_us_fiscal_year()
	fiscal_year_start = fy_info["fiscal_year_start"]
	fiscal_year_end = fy_info["fiscal_year_end"]
	doc = frappe.get_all(
		"Leave Allocation",
		filters={
			"employee": employee,
			"from_date": ["between", [fiscal_year_start, fiscal_year_end]],
			"docstatus": 1,
		},
		fields=["name", "new_leaves_allocated"],
	)
	if doc:
		existing_amount = float(doc[0]["new_leaves_allocated"])
		if float(amount) != existing_amount:
			return {
				"status": "exists_different_amount",
				"existing_amount": existing_amount,
				"employee": employee,
				"new_amount": float(amount),
			}

		return {"status": "exists_same_amount", "employee": employee}

	return {"employee": employee, "amount": float(amount)}


@frappe.whitelist()
def update_allocated_leaves(
	employee: str,
	amount: float,
	reason: str,
	existing_amount: float,
):
	fy_info = get_us_fiscal_year()
	fiscal_year_start = fy_info["fiscal_year_start"]
	fiscal_year_end = fy_info["fiscal_year_end"]
	doc = frappe.get_all(
		"Leave Allocation",
		filters={
			"employee": employee,
			"from_date": ["between", [fiscal_year_start, fiscal_year_end]],
			"docstatus": 1,
		},
		fields=["name", "new_leaves_allocated"],
	)

	if doc:
		existing_amount = float(existing_amount)
		existing_doc = frappe.get_doc("Leave Allocation", doc[0].name)
		existing_doc.new_leaves_allocated = amount
		existing_doc.custom_allocation_change_reason = reason
		existing_doc.custom_previous_leaves_allocated = existing_amount
		existing_doc.save()
	return {"status": "updated", "existing_amount": existing_amount}


@frappe.whitelist()
def get_us_fiscal_year():
	from frappe.utils import getdate, now_datetime

	today = now_datetime().date()
	year = today.year
	if today.month < 10:
		start_year = year - 1
		end_year = year
	else:
		start_year = year
		end_year = year + 1

	fiscal_year_start = getdate(f"{start_year}-10-01")
	fiscal_year_end = getdate(f"{end_year}-09-30")
	return {
		"fiscal_year": f"{start_year}-{end_year}",
		"fiscal_year_start": fiscal_year_start,
		"fiscal_year_end": fiscal_year_end,
	}
